I"‹<p>Python has many built-in exceptions that are raised when your program encounters an error (something in the program goes wrong). When these exceptions occur, the Python interpreter stops the current process and passes it to the calling process until it is handled. If not handled, the program will crash.</p>

<p>Error handling in python is mainly handled by <code class="language-plaintext highlighter-rouge">try</code>, <code class="language-plaintext highlighter-rouge">except</code> and <code class="language-plaintext highlighter-rouge">finally</code></p>

<ul>
  <li>
    <p>The <code class="language-plaintext highlighter-rouge">try</code> block lets you test a block of code for errors.</p>
  </li>
  <li>
    <p>The <code class="language-plaintext highlighter-rouge">except</code> block lets you handle the error.</p>
  </li>
  <li>
    <p>The <code class="language-plaintext highlighter-rouge">finally</code> block lets you execute code, regardless of the result of the <code class="language-plaintext highlighter-rouge">try</code> and <code class="language-plaintext highlighter-rouge">except</code> blocks.</p>
  </li>
</ul>

<pre><code class="language-python3">
#The try block will generate a NameError, because x is not defined:

try:
  print(x)
except:
  print("Variable x is not defined")

# output

Variable x is not defined

</code></pre>

<ul>
  <li>We can define as many exception blocks as you want, i.e if we want to execute a special block of code for a special kind of error.</li>
</ul>

<pre><code class="language-python3">
try:
  print(x)

except NameError:

# block for NameError

  print("Variable x is not defined")

except:

# block for Other exception

  print("Something else went wrong")

</code></pre>

<h1 id="else">Else</h1>

<p>We can use the <code class="language-plaintext highlighter-rouge">else</code> keyword to define a block of code to be executed if no errors were raised.</p>

<pre><code class="language-print3">
try:
  print("Hello")
except:
  print("Something went wrong")
else:
  print("Nothing went wrong")


# Output

Hello
Nothing went wrong
</code></pre>

<h1 id="finally">Finally</h1>

<p>The <code class="language-plaintext highlighter-rouge">finally</code> block, if specified, will be executed regardless if the try block raises an error or not.</p>

<pre><code class="language-python3">
try:
  print(x)
except:
  print("Something went wrong")
finally:
  print("The 'try except' is finished")

#Output
Something went wrong
The 'try except' is finished
</code></pre>

<h1 id="raise-an-exception">Raise an exception</h1>

<p>As a Python developer you can choose to throw an exception if a condition occurs. To throw (or raise) an exception, use the raise keyword.</p>

<pre><code class="language-python3">
# Raise an error and stop the program if x is lower than 0:

x = -1

if x &lt; 0:
  raise Exception("Sorry, no numbers below zero")

</code></pre>

<p><img src="/beg" alt="Exceptions" /></p>
:ET